File: tools/rendre_en.py
# -*- coding: utf-8 -*-
"""Rend le balisage des pages produit EN ANGLAIS, au moment du build.

Le probleme, tel qu'il se voit vraiment :

  Les quatre pages produit naissent en francais. Le balisage porte le
  francais, un voile cache la page, `applyLang()` la traduit, le voile se
  leve. Un visiteur avec JavaScript ne voit jamais le francais.

  Mais `<html lang="fr">` restait ecrit, et le corps aussi. Un moteur
  d'indexation lit du francais. Un lecteur d'ecran annonce du francais et le
  prononce a la francaise. Et quiconque perd le script — reseau coupe en
  cours de chargement, extension trop zelee, navigateur ancien — reste devant
  un voile qui ne se levera pas.

CE QUE FAIT CE MODULE. Il execute le script DE LA PAGE (voir rendre_en.mjs)
avec un DOM d'enregistrement, note quel element recoit quel texte anglais,
puis pose ces textes dans le balisage. La traduction n'est donc pas devinee :
elle vient de la page elle-meme, par le meme chemin qu'au navigateur.

⚠️ ON NE REANALYSE PAS LE DOCUMENT. Les remplacements sont chirurgicaux, sur
la chaine. Une reserialisation par un analyseur HTML toucherait aux scripts en
ligne — et leurs empreintes SHA-256 sont ce sur quoi repose la CSP.
"""
import io
import json
import os
import re
import subprocess
import sys

RACINE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
VIDES = {"input", "img", "br", "hr", "meta", "link", "source", "area", "col"}


def zones_script(html):
    """Les intervalles occupes par du script : on n'y cherche aucun id."""
    return [(m.start(), m.end())
            for m in re.finditer(r"<script\b[^>]*>.*?</script>", html, re.S | re.I)]


def dans_script(zones, i):
    return any(a <= i < b for a, b in zones)


def echapper(v):
    return v.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def interieur(html, ident, zones):
    """(debut, fin, nom_de_balise) du contenu de l'element portant cet id."""
    motif = re.compile(r'<([a-zA-Z][\w-]*)\b[^>]*?\sid="%s"' % re.escape(ident))
    for m in motif.finditer(html):
        if dans_script(zones, m.start()):
            continue
        nom = m.group(1).lower()
        if nom in VIDES:
            return None
        i = html.find(">", m.end())
        if i < 0 or html[i - 1] == "/":
            return None
        debut, prof, j = i + 1, 1, i + 1
        bal = re.compile(r"<(/?)%s\b" % re.escape(nom), re.I)
        while prof:
            k = bal.search(html, j)
            if not k:
                return None
            if k.group(1):
                prof -= 1
                if prof == 0:
                    return debut, k.start(), nom
            else:
                f = html.find(">", k.end())
                if f < 0:
                    return None
                if html[f - 1] != "/":
                    prof += 1
            j = k.end()
    return None


def poser(html, ident, mode, valeur):
    zones = zones_script(html)
    p = interieur(html, ident, zones)
    if not p:
        return html, False
    debut, fin, _ = p
    v = valeur if mode == "innerHTML" else echapper(valeur)
    return html[:debut] + v + html[fin:], True


def poser_options(html, ident, rangs):
    zones = zones_script(html)
    p = interieur(html, ident, zones)
    if not p:
        return html, 0
    debut, fin, _ = p
    bloc = html[debut:fin]
    n = 0

    compteur = {"i": -1}

    def une(m):
        compteur["i"] += 1
        t = rangs.get(str(compteur["i"]))
        if t is None:
            return m.group(0)
        return m.group(1) + echapper(t) + m.group(3)

    bloc2, _c = re.subn(r"(<option\b[^>]*>)(.*?)(</option>)", une, bloc, flags=re.S | re.I)
    n = sum(1 for k in rangs if int(k) <= compteur["i"])
    return html[:debut] + bloc2 + html[fin:], n


def poser_titre(html, titre, descr):
    faits = 0
    if titre:
        h, ok = re.subn(r"(<title\b[^>]*>).*?(</title>)",
                        lambda m: m.group(1) + echapper(titre) + m.group(2),
                        html, count=1, flags=re.S | re.I)
        html, faits = h, faits + ok
    if descr:
        def rempl(m):
            return re.sub(r'content="[^"]*"', 'content="%s"' % descr.replace('"', "&quot;"),
                          m.group(0))
        h, ok = re.subn(r'<meta\s+name="description"[^>]*>', rempl, html, count=1, flags=re.I)
        html, faits = h, faits + ok
    return html, faits


def poser_langue(html):
    """`lang="en"`, replis anglais, et le voile qui change de camp."""
    n = 0
    html, k = re.subn(r'<html lang="[a-z-]+"', '<html lang="en"', html, count=1)
    n += k
    # Le repli de `lang()` : il restait au francais DANS LE CORPS de la page,
    # alors que l'entete avait deja bascule a l'anglais. Une langue absente du
    # dictionnaire — de l'allemand, du portugais — retombait donc en francais.
    html, k = re.subn(r'(document\.documentElement\.lang\|\|localStorage\.getItem\("atmart_lang"\)\|\|)"fr"',
                      r'\1"en"', html)
    n += k
    html, k = re.subn(r'(return L\[l\]\?l:)"fr";', r'\1"en";', html)
    n += k
    # ⚠️ LE VOILE CHANGEAIT DE CAMP AVEC LE BALISAGE. Il cachait la page tant
    # qu'on n'etait PAS en francais, parce que le francais etait ecrit dans le
    # HTML. Le balisage est desormais anglais : c'est l'anglais qui n'a plus
    # besoin d'attendre, et tout le reste qui attend.
    html, k = re.subn(r'if\(l!=="fr"\)\{d\.className\+=" i18n-wait"',
                      'if(l!=="en"){d.className+=" i18n-wait"', html)
    n += k
    return html, n


def rendre(chemin, langue="en", bavard=True):
    mjs = os.path.join(RACINE, "tools", "rendre_en.mjs")
    r = subprocess.run(["node", mjs, chemin, langue],
                       capture_output=True, text=True, encoding="utf-8")
    if r.returncode != 0:
        raise SystemExit("rendre_en.mjs a echoue sur %s :\n%s" % (chemin, r.stderr[-2000:]))
    d = json.loads(r.stdout)
    if d["soucis"]:
        # Un script qui meurt en route, c'est une page a moitie traduite. On
        # refuse de publier plutot que de livrer un melange de langues.
        raise SystemExit("%s : le script de la page a echoue — %s"
                         % (os.path.basename(chemin), " · ".join(d["soucis"])))

    html = io.open(chemin, encoding="utf-8").read()

    def presents(t):
        z = zones_script(t)
        return {m.group(1) for m in re.finditer(r'\sid="([^"]+)"', t)
                if not dans_script(z, m.start())}

    avant = presents(html)

    poses = ratees = 0
    for ident, e in d["ecrits"].items():
        html, ok = poser(html, ident, e["mode"], e["v"])
        poses += ok
        ratees += not ok

    # ⚠️ UN PARENT PEUT AVALER SON ENFANT. Si applyLang() ecrit l'innerHTML
    # d'un element QUI EN CONTIENT un autre qu'elle traduit aussi, poser le
    # parent efface l'identifiant de l'enfant — le texte reste juste en
    # anglais, mais la page ne sait plus le retraduire, et passer en kreyol
    # laisse cette ligne-la en anglais. Le defaut serait invisible au build et
    # ne se verrait qu'en changeant de langue, sur une ligne parmi cinquante.
    disparus = sorted(avant - presents(html))
    if disparus:
        raise SystemExit(
            "%s : %d identifiant(s) efface(s) en posant l'anglais — %s.\n"
            "Un element traduit en contient un autre : la page ne pourra plus "
            "retraduire l'enfant quand le visiteur changera de langue."
            % (os.path.basename(chemin), len(disparus), " ".join(disparus)))
    opts = 0
    for ident, rangs in d["options"].items():
        html, k = poser_options(html, ident, rangs)
        opts += k
    html, t = poser_titre(html, d["titre"], d["description"])
    html, lg = poser_langue(html)

    tmp = chemin + ".tmp"
    io.open(tmp, "w", encoding="utf-8", newline="\n").write(html)
    os.replace(tmp, chemin)
    if bavard:
        print("     %-14s %3d pose(s) · %2d option(s) · %d titre/descr · %d langue"
              % (os.path.basename(chemin), poses, opts, t, lg)
              + ("  !! %d id(s) introuvable(s) dans le balisage" % ratees if ratees else ""))
    return poses, ratees


if __name__ == "__main__":
    cibles = sys.argv[1:] or ["wout.html", "setdi.html", "vivye.html", "anplwaye.html"]
    total = 0
    for c in cibles:
        _, r = rendre(os.path.join(RACINE, c))
        total += r
    sys.exit(0)

File: tools/rendre_en.mjs
// Enregistreur de langue : rejoue applyLang() de la page, en anglais, hors
// navigateur, et dit quel element recoit quel texte.
//
// POURQUOI CE DETOUR. Les quatre pages produit naissent en francais : le
// balisage porte le francais, un voile cache la page, puis applyLang() la
// traduit et le voile se leve. Un visiteur avec JavaScript ne voit jamais le
// francais — mais un moteur d'indexation, un lecteur d'ecran et quiconque
// perd le script voient une page francaise annoncee <html lang="fr">.
//
// On ne peut pas deviner la traduction par expression reguliere : chaque page
// branche ses cles a sa facon (textContent, innerHTML, setTxt, setH, et des
// gardes conditionnelles). On execute donc le script DE LA PAGE avec un DOM
// d'enregistrement : ce n'est pas moi qui decide quel texte va ou, c'est la
// page elle-meme. Ce qu'elle ecrit, on le note ; on l'appliquera au balisage.
//
// Usage : node tools/rendre_en.mjs <fichier.html> <langue>
import fs from "node:fs";
import vm from "node:vm";

const [, , fichier, langue = "en"] = process.argv;
const html = fs.readFileSync(fichier, "utf8");

const ecrits = Object.create(null); // id -> {mode, v}
const options = Object.create(null); // id de <select> -> {rang: texte}
const soucis = [];

// ⚠️ LES <option> D'UNE LISTE DEROULANTE SE TRADUISENT PAR RANG, PAS PAR
// IDENTIFIANT : la page ecrit `select.options[0].text = t.a0`. Sans ce relais,
// `options` etait un tableau vide, `options[0]` valait undefined, et le script
// de vivye.html mourait AVANT de poser le titre de l'onglet — d'ou une page
// dont seule une moitie se traduisait.
function listeOptions(idSelect) {
  const rangs = new Map();
  return new Proxy({ length: 8 }, {
    get(o, p) {
      if (p === "length") return o.length;
      if (typeof p === "symbol" || !/^\d+$/.test(String(p))) return undefined;
      const r = Number(p);
      if (!rangs.has(r)) {
        rangs.set(r, new Proxy({ value: "", selected: false, text: "" }, {
          get(t2, p2) { return p2 in t2 ? t2[p2] : undefined; },
          set(t2, p2, v) {
            if (idSelect && (p2 === "text" || p2 === "textContent" || p2 === "label")) {
              (options[idSelect] ||= {})[r] = String(v);
            }
            t2[p2] = v;
            return true;
          },
        }));
      }
      return rangs.get(r);
    },
  });
}

const RIEN = () => undefined;
const methodes = {
  addEventListener: RIEN, removeEventListener: RIEN, dispatchEvent: RIEN,
  appendChild: RIEN, removeChild: RIEN, insertBefore: RIEN, remove: RIEN,
  append: RIEN, prepend: RIEN, replaceChildren: RIEN, cloneNode: RIEN,
  focus: RIEN, blur: RIEN, click: RIEN, submit: RIEN, reset: RIEN,
  setAttribute: RIEN, removeAttribute: RIEN, scrollIntoView: RIEN,
  insertAdjacentHTML: RIEN, insertAdjacentElement: RIEN, animate: RIEN,
  play: RIEN, pause: RIEN, load: RIEN, select: RIEN, setSelectionRange: RIEN,
  getAttribute: () => null, hasAttribute: () => false, closest: () => null,
  matches: () => false, contains: () => false,
  getBoundingClientRect: () => ({ top: 0, left: 0, width: 0, height: 0, bottom: 0, right: 0 }),
};

function element(id) {
  const t = {
    id: id || "", tagName: "DIV", nodeType: 1,
    dataset: {}, style: {}, classList: {
      add: RIEN, remove: RIEN, toggle: RIEN, contains: () => false, replace: RIEN,
    },
    children: [], childNodes: [], parentNode: null, firstChild: null,
    nextElementSibling: null, previousElementSibling: null,
    value: "", checked: false, disabled: false, selectedIndex: 0,
    options: listeOptions(id),
    textContent: "", innerHTML: "", innerText: "", outerHTML: "",
    className: "", href: "", src: "", hidden: false, files: [],
    offsetWidth: 0, offsetHeight: 0, scrollTop: 0, scrollHeight: 0,
    ...methodes,
    querySelector: () => element(""),
    querySelectorAll: () => [],
    getElementsByTagName: () => [],
    getElementsByClassName: () => [],
  };
  return new Proxy(t, {
    get(o, p) {
      if (p in o) return o[p];
      if (typeof p === "symbol") return undefined;
      return undefined;
    },
    set(o, p, v) {
      // ⚠️ C'EST ICI QUE TOUT SE JOUE. On ne note que les elements qui ont un
      // identifiant : ce sont les seuls qu'on saura retrouver dans le balisage.
      if (id && (p === "textContent" || p === "innerHTML")) {
        ecrits[id] = { mode: p, v: String(v) };
      }
      o[p] = v;
      return true;
    },
  });
}

const cache = new Map();
function parId(id) {
  if (!cache.has(id)) cache.set(id, element(id));
  return cache.get(id);
}

const racine = element("");
racine.lang = langue;
racine.className = "";

let titre = "";
let description = "";

const metaDescr = element("");
metaDescr.setAttribute = (n, v) => { if (n === "content") description = String(v); };

const document = new Proxy({
  documentElement: racine,
  body: element(""),
  head: element(""),
  readyState: "loading",
  cookie: "",
  getElementById: (id) => parId(String(id)),
  querySelector: (s) => (/meta\[name="description"\]/.test(String(s)) ? metaDescr : element("")),
  querySelectorAll: () => [],
  getElementsByTagName: () => [],
  getElementsByClassName: () => [],
  createElement: () => element(""),
  createTextNode: () => element(""),
  createDocumentFragment: () => element(""),
  addEventListener: RIEN, removeEventListener: RIEN,
  execCommand: RIEN, write: RIEN, close: RIEN,
  activeElement: null,
}, {
  get(o, p) {
    if (p === "title") return titre;
    if (p in o) return o[p];
    return undefined;
  },
  set(o, p, v) {
    if (p === "title") { titre = String(v); return true; }
    o[p] = v;
    return true;
  },
});

const stockage = {
  getItem: (k) => (k === "atmart_lang" ? langue : null),
  setItem: RIEN, removeItem: RIEN, clear: RIEN, key: () => null, length: 0,
};

const fenetre = {
  document,
  localStorage: stockage, sessionStorage: stockage,
  location: { href: "https://driver360.atmart.ltd/", search: "", hash: "", pathname: "/", origin: "https://driver360.atmart.ltd", protocol: "https:", host: "driver360.atmart.ltd", reload: RIEN, replace: RIEN, assign: RIEN },
  navigator: { language: langue, languages: [langue], userAgent: "node", serviceWorker: { register: RIEN }, clipboard: { writeText: RIEN }, mediaDevices: {}, onLine: true },
  history: { pushState: RIEN, replaceState: RIEN, back: RIEN },
  // ⚠️ AUCUN MINUTEUR NE PART. On veut l'etat de la page au premier affichage,
  // pas ce qu'elle deviendra apres un delai — et surtout aucune attente.
  setTimeout: RIEN, setInterval: RIEN, clearTimeout: RIEN, clearInterval: RIEN,
  requestAnimationFrame: RIEN, cancelAnimationFrame: RIEN,
  fetch: () => new Promise(() => {}),         // ne se resout jamais : rien ne suit
  MutationObserver: class { observe() {} disconnect() {} },
  IntersectionObserver: class { observe() {} disconnect() {} },
  ResizeObserver: class { observe() {} disconnect() {} },
  speechSynthesis: { speak: RIEN, cancel: RIEN, getVoices: () => [] },
  SpeechSynthesisUtterance: class {},
  matchMedia: () => ({ matches: false, addEventListener: RIEN, addListener: RIEN }),
  addEventListener: RIEN, removeEventListener: RIEN,
  print: RIEN, alert: RIEN, confirm: () => false, prompt: () => null,
  open: RIEN, scrollTo: RIEN, getComputedStyle: () => ({ getPropertyValue: () => "" }),
  URLSearchParams, URL, JSON, Math, Date, console,
  crypto: { getRandomValues: (a) => a, randomUUID: () => "0" },
  innerWidth: 1280, innerHeight: 800, devicePixelRatio: 1,
};
fenetre.window = fenetre;
fenetre.self = fenetre;
fenetre.globalThis = fenetre;

const contexte = vm.createContext(fenetre);

// Les scripts en ligne, dans l'ordre du document.
const scripts = [...html.matchAll(/<script\b(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/gi)]
  .map((m) => m[1]);

for (const [i, code] of scripts.entries()) {
  try {
    new vm.Script(code, { filename: `${fichier}#${i}` }).runInContext(contexte, { timeout: 5000 });
  } catch (e) {
    // Un script qui echoue n'annule pas les autres : on note et on continue.
    soucis.push(`script #${i} : ${e.message}`);
  }
  // La langue ne doit jamais deriver : le script d'entete la recalcule.
  racine.lang = langue;
}

process.stdout.write(JSON.stringify({
  langue, titre, description, ecrits, options, soucis,
  nb: Object.keys(ecrits).length,
}, null, 1));
